extensions/amp-ad-network-google-ads-impl/0.1/url-builder.js

```
@@ -0,0 +1,141 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Builds a URL from query parameters, truncating to a maximum length if
+ * necessary.
+ * @param {string} baseUrl scheme, domain, and path for the URL.
+ * @param {!Array<!QueryParameter>} requiredQueryParams query parameters that
+ *     must appear in the URL if available and cannot be truncated.
+ * @param {?number=} opt_maxLength length to truncate the URL to if necessary.
+ * @param {?Array<!QueryParameter>=} opt_optionalQueryParams query parameters to
+ *     try to include in the URL, but that may be truncated if necessary.  These
+ *     are prioritized in order.
+ * @param {?QueryParameter=} opt_truncationQueryParam query parameter that must
+ *     be included in the URL iff any of the optional query parameters were
+ *     truncated.
+ * @return {string} the fully constructed URL.
+ */
+export function buildUrl(
+    baseUrl, requiredQueryParams, opt_maxLength, opt_optionalQueryParams,
+    opt_truncationQueryParam) {
+  const encodedParams = requiredQueryParams
+      .map(param => param.encode_())
+      .filter(encoded => encoded != null);
+  const optionalQueryParams = opt_optionalQueryParams || [];
+  const encodedParamsWithoutTruncation = encodedParams.concat(
+      optionalQueryParams
+          .map(param => param.encode_())
+          .filter(encoded => encoded != null));
+  if (opt_maxLength == null ||
+      baseUrl.length + encodedParamsWithoutTruncation.length +
+      totalLength(encodedParamsWithoutTruncation) <= opt_maxLength) {
```

My original reason for not doing this was out of a desire to distinguish between parameters that are expected to be included in their entirety vs. those that can safely be truncated. As you observed in the comment above this one, in the current implementation there's not much of a difference because I don't handle the case where the base URL and required params by themselves exceed the maximum length. If we want to throw an error or log a warning if the required params don't all fit, then the distinction needs to be maintained; on the other hand, if we're okay with truncating arbitrary parameters then we can do it the simpler way. (Of course, if we put the fixed-length parameters first then those won't be truncated; the problem is that some of the critical parameters are user-supplied and so we don't know that they won't be arbitrarily long.)